### PR TITLE
make: cleanup and centralise DEVELHELP setting

### PR DESCRIPTION
**Pull request by smlng**
This PR make the usage of DEVELHELP more consistent and easier to configure (IMHO 😄), i.e.,
its no longer required to explicitly set `CFLAGS += -DDEVELHELP` in Makefiles but rather use `DEVELHELP=1` either in Makefile or as env variable on command line; specifically for the latter its simpler.

Further this PR make usage of DEVELHELP more consistent, as follows:

- remove setting of DEVELHELP for all examples, can be optionally activated as described above
- enable DEVELHELP for all test by default, in `tests/Makefile.test_common`

### CODE REVIEW

**Comment by smlng** [comment]
requires #8030

File: tests/openthread/Makefile [hunk 1]
```diff
@@ -8,11 +8,6 @@ BOARD_WHITELIST := samr21-xpro iotlab-m3 fox iotlab-a8-m3
 # This has to be the absolute path to the RIOT base directory:
 RIOTBASE ?= $(CURDIR)/../..
 
-# Comment this out to disable code in RIOT that does safety checking
-# which is not needed in a production environment but helps in the
-# development process:
-CFLAGS += -DDEVELHELP -Wall
```
Comment (cladmi) [inline tests/openthread/Makefile]: `-Wall` got removed at the same time here

File: Makefile.include [hunk 2]
```diff
@@ -111,6 +111,14 @@ endif
 
 QQ=
 
+# Set this to 1 to enable code in RIOT that does safety checking
+# which is not needed in a production environment but helps in the
+# development process:
```
Comment (cladmi) [inline Makefile.include]: You could also add that it enables the `DEBUG_PRINT` macro. https://github.com/RIOT-OS/RIOT/blob/2017.10-branch/core/include/debug.h#L46

**Comment by kaspar030** [comment]
I like!

> remove setting of DEVELHELP for all examples, can be optionally activated as described above

I think this is a rather big change and should maybe be kept out of this PR.

**Comment by smlng** [comment]
> I think this is a rather big change and should maybe be kept out of this PR.

mhm, it wasn't used consistently and also makes not much sense to have DEVELHELP active for examples by default (IMHO). Iff we want to have develhelp active nevertheless I would still adapt it here to set `DEVELHELP = 1` in respective Makefile of the examples, instead of keeping the `CFLAG` variant

**Comment by kaspar030** [comment]
> would still adapt it here to set DEVELHELP = 1 in respective Makefile of the examples

I think that would be the way to go!

**Comment by smlng** [comment]
oh great, Murdock reveals failed assertion in gnrc:

```
main(): This is RIOT! (Version: buildtest)
...................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.c:772 => 0x8049ba9
*** RIOT kernel panic:
FAILED ASSERTION.
```

and further 

```
gnrc_sixlowpan_nd_router.c: In function 'gnrc_sixlowpan_nd_router_abr_rem_prf':
gnrc_sixlowpan_nd_router.c:330:62: error: unused parameter 'iface' [-Werror=unused-parameter]
                                           gnrc_ipv6_netif_t *iface, gnrc_ipv6_netif_addr_t *prefix)
                                                              ^
```

so my PR will be blocked by this (IMHO nonsense) GNRC merge embargo 😞

**Comment by miri64** [comment]
Can we let `DEVELHELP` activated on Murdock? Or is that not desirable for future `size` monitoring?

**Comment by miri64** [comment]
(otherwise: should be pretty straight forward by setting `DEVELHELP=1` in `.murdock`, right?

**Comment by kaspar030** [comment]
> Or is that not desirable for future size monitoring?

I'd say murdock shouldn't compile differently from what the Makefile says, if possible. Also, for size monitoring, DEVELHELP shouldn't be set. (At least for tests/minimal, examples/hello-world, examples/gnrc_minimal ...)

**Comment by smlng** [comment]
regarding murdock, I think we should compile everything with and without DEVELHELP=1, as it checks out differently - at least for the nightlies we should do that.

**Comment by kaspar030** [comment]
> enable DEVELHELP for all test by default, in tests/Makefile.test_common

Can you disable it for tests/minimal?

**Comment by kaspar030** [comment]
> DEVELHELP shouldn't be set. (At least for tests/minimal, examples/hello-world, examples/gnrc_minimal ...)

Sorry, I meant, "shouldn't be changed"

**Comment by smlng** [comment]
@kaspar030 I excluded `tests/minimal` from DEVELHELP

**Comment by smlng** [comment]
I still DEVELHELP doesn't make sense for examples by default

**Comment by cladmi** [comment]
I would prefer separating this in two PRs:
* Introduce the new configuration mechanism
* Sanitize the usage of DEVELHELP

File: tests/openthread/Makefile (hunk 1 above)
Comment (smlng) [reply tests/openthread/Makefile]: IMHO wasn't appropriate place to activate that here anyway, use `WERROR` (see e.g. #7919) instead

**Comment by smlng** [comment]
> I would prefer this separated in two PRs:

I don't see the usefulness of such a split, separate commits: certainly, but PRs: what for. Introducing the changed config ability makes (IMHO) no sense without using it too, where ever appropriate.

**Comment by kaspar030** [comment]
> I don't see the usefulness of such a split, separate commits: certainly, but PRs: what for.

When testing multiple changes at the same time, the testing effort is the product of the effort needed to test each individual change. Same goes with the discussion.

**Comment by smlng** [comment]
factored out changes as requested, this now only introduces the new develhelp switch.

PR depends on #8030, #8079, #8080, and #8081

**Comment by kaspar030** [comment]
> PR depends on #8030.

Sorry, but why?

**Comment by kaspar030** [comment]
> only introduces the new develhelp switch.

I'm sorry to be picky on this, but the switch also has to be used within the makefiles, instead of setting CFLAGS directly.

**Comment by smlng** [comment]
> PR depends on #8030.
Sorry, but why?

Right, this one doesn't need it but the #8080 does, otherwise `tests/pkg_libcoap` fails. Will drop here.

**Comment by smlng** [comment]
> only introduces the new develhelp switch.
I'm sorry to be picky on this, but the switch also has to be used within the makefiles, instead of setting CFLAGS directly.

I don't understand, you said the change should be factored out, hence #8080 and #8081. Or what do you mean?

**Comment by smlng** [comment]
good point @miri64, will do!

**Comment by smlng** [comment]
@miri64 the doc is rather generic already, I don't see any point in fixing it - but could be extended?!

**Comment by miri64** [comment]
> I don't see any point in fixing it - but could be extended?!

Extension was what I was asking for. At least with regards to this PR. Something like the sentence "Can be set by setting environment variable `DEVELHELP` is set to 1 at compile time." would already satisfy me.

File: doc.txt [hunk 3]
```diff
@@ -9,8 +9,11 @@
 
 /**
  * @def DEVELHELP
- * @brief   This global macro activates some behavior that helps you while
- *          developing but is otherwise optimized out.
+ * @brief   This global macro activates functionality to help developers.
+ *
+ *          Additional code parts such as (extensive) debug output and sanity
+ *          checks using assertions. To activate set environment variable
+ *          `DEVELHELP=1`, or disable (if on by default) with `DEVELHELP=0`.
```
Comment (miri64) [inline doc.txt]: the "if" looks a little bit out of place here.

**Comment by smlng** [comment]
@miri64 changed wording, better?

**Pull request merged**